File: chesstopia-backend/src/test/java/io/chesstopia/backend/game/domain/PositionTest.java
package io.chesstopia.backend.game.domain;

import org.junit.jupiter.api.Test;
import java.util.Map;
import static org.assertj.core.api.Assertions.*;

class PositionTest {

    @Test
    void pieceAtFindetUndVermisstFiguren() {
        // ARRANGE
        var pos = new Position(
            Map.of(new Square(File.E, Rank.ONE), new Piece(PieceType.KING, Color.WHITE)),
            Color.WHITE, CastlingRights.all(), null, 0, 1);

        // ACT & ASSERTIONS
        assertThat(pos.pieceAt(new Square(File.E, Rank.ONE)))
            .contains(new Piece(PieceType.KING, Color.WHITE));
        assertThat(pos.pieceAt(new Square(File.E, Rank.TWO))).isEmpty();
    }

    @Test
    void diePieceMapIstNachAussenUnveraenderlich() {
        // ARRANGE
        var pieces = new java.util.HashMap<Square, Piece>();
        pieces.put(new Square(File.A, Rank.ONE), new Piece(PieceType.ROOK, Color.WHITE));
        var pos = new Position(pieces, Color.WHITE, CastlingRights.all(), null, 0, 1);

        // ACT
        // Nachträgliche Änderung an der übergebenen Map darf die Position nicht berühren.
        pieces.clear();

        // ASSERTIONS
        assertThatThrownBy(() -> pos.pieces().clear())
            .isInstanceOf(UnsupportedOperationException.class);
        assertThat(pos.pieces()).hasSize(1);
    }
}

File: chess-engine/src/commonTest/kotlin/io/chesstopia/engine/PositionTest.kt
package io.chesstopia.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class PositionTest {

    @Test
    fun `die Startstellung besetzt genau 32 Felder`() {
        // ACT
        val p = standardStartPosition()

        // ASSERTIONS
        assertEquals(32, p.board.size)
        assertEquals(Piece(PieceType.ROOK, Color.BLACK), p.pieceAt(Square(File.A, Rank.EIGHT)))
        assertEquals(Piece(PieceType.KING, Color.WHITE), p.pieceAt(Square(File.E, Rank.ONE)))
        assertNull(p.pieceAt(Square(File.E, Rank.FOUR)))
    }

    @Test
    fun `die Startstellung hat Weiß am Zug, alle Rochaderechte, keinen Halbzug`() {
        // ACT
        val p = standardStartPosition()

        // ASSERTIONS
        assertEquals(Color.WHITE, p.sideToMove)
        assertEquals(CastlingRights.ALL, p.castlingRights)
        assertNull(p.enPassantTarget)
        assertEquals(0, p.halfmoveClock)
        assertEquals(1, p.fullmoveNumber)
    }

    @Test
    fun `zwei gleich aufgebaute Stellungen sind gleich`() {
        // ACT & ASSERTIONS
        assertEquals(standardStartPosition(), standardStartPosition())
    }

    @Test
    fun `die Reihenfolge der Feldliste ändert die Gleichheit nicht`() {
        // ARRANGE
        val a = position(
            Square(File.E, Rank.TWO) to Piece(PieceType.PAWN, Color.WHITE),
            Square(File.D, Rank.FOUR) to Piece(PieceType.KNIGHT, Color.BLACK),
        )
        val b = position(
            Square(File.D, Rank.FOUR) to Piece(PieceType.KNIGHT, Color.BLACK),
            Square(File.E, Rank.TWO) to Piece(PieceType.PAWN, Color.WHITE),
        )

        // ACT & ASSERTIONS
        assertEquals(a, b)
        assertEquals(a.hashCode(), b.hashCode())
    }
}
